src/utils/ucc_coll_utils.c

```diff
@@ -330,55 +329,204 @@ ucc_coll_args_is_rooted(const ucc_base_coll_args_t *bargs)
+        return 1;
+    }
+    return 0;
+}
+
+#define COLL_ARGS_HEADER_STR_MAX_SIZE 32
+void ucc_coll_args_str(const ucc_coll_args_t *args, ucc_rank_t trank,
+                       ucc_rank_t tsize, char *str, size_t len)
 {
-    const ucc_base_coll_args_t *args  = &task->bargs;
-    ucc_team_t                *team  = args->team;
-    ucc_coll_type_t            ct    = args->args.coll_type;
-    size_t                     left  = len;
-    char                       tmp[64];
-
-    ucc_snprintf_safe(str, left, "req %p, seq_num %u, %s, team_id %u, size %u, "
-                      "rank %u, ctx_rank %u: %s %s inplace=%u",
-                      task, task->seq_num,
-                      task->team->context->lib->log_component.name,
-                      team->id, team->size, team->rank,
-                      ucc_ep_map_eval(team->ctx_map, team->rank),
-                      ucc_coll_type_str(ct),
-                      ucc_mem_type_str(ucc_coll_args_mem_type(&args->args,
-                                                              team->rank)),
-                      UCC_IS_INPLACE(args->args));
-
-    if (ucc_coll_args_is_rooted(args)) {
-        ucc_snprintf_safe(tmp, sizeof(tmp), " root=%u", (ucc_rank_t)args->args.root);
-        left = len - strlen(str);
-        strncat(str, tmp, left);
+    ucc_coll_type_t        ct                                 = args->coll_type;
+    ucc_rank_t             root                               = args->root;
+    ucc_coll_buffer_info_t src_info                           = {0};
+    ucc_coll_buffer_info_t dst_info                           = {0};
+    char                   hdr[COLL_ARGS_HEADER_STR_MAX_SIZE] = "";
+    char tmp[32];
+    size_t count;
+    int left, has_src, has_dst;
+
+    ucc_snprintf_safe(hdr, COLL_ARGS_HEADER_STR_MAX_SIZE, "%s",
+                      ucc_coll_type_str(ct));
+    if (ucc_coll_args_is_reduction(ct)) {
+        ucc_snprintf_safe(tmp, sizeof(tmp), " %s",
+                          ucc_reduction_op_str(args->op));
+        left = COLL_ARGS_HEADER_STR_MAX_SIZE - strlen(hdr);
+        strncat(hdr, tmp, left);
+    }
+
+    if (UCC_IS_INPLACE(*args)) {
+        ucc_snprintf_safe(tmp, sizeof(tmp), " inplace");
+        left = COLL_ARGS_HEADER_STR_MAX_SIZE - strlen(hdr);
+        strncat(hdr, tmp, left);
+    }
+
+    if (ucc_coll_args_is_rooted(ct)) {
+        ucc_snprintf_safe(tmp, sizeof(tmp), " root %u", root);
+        left = COLL_ARGS_HEADER_STR_MAX_SIZE - strlen(hdr);
+        strncat(hdr, tmp, left);
+    }
+
+    has_src = has_dst = 0;
+    switch (ct) {
+    case UCC_COLL_TYPE_ALLGATHER:
+    case UCC_COLL_TYPE_ALLREDUCE:
+    case UCC_COLL_TYPE_ALLTOALL:
+    case UCC_COLL_TYPE_REDUCE_SCATTER:
+        dst_info = args->dst.info;
+        has_dst = 1;
+        if (!UCC_IS_INPLACE(*args)) {
+            src_info = args->src.info;
+            has_src = 1;
+        }
+        break;
+    case UCC_COLL_TYPE_ALLGATHERV:
+    case UCC_COLL_TYPE_REDUCE_SCATTERV:
```

Comment:
maybe move thise entire blob into separate util function:
```
static inline void ucc_coll_args_get_info(const ucc_coll_args_t *args, ucc_coll_buffer_info_t *s_info, 
ucc_coll_buffer_info_t *d_info) {
...
}
```
it can set s_info.buffer or r_info.buffer to NULL that would indicate the same as has_src/has_dst flags.

Reply:
s_info.buffer == NULL is actually valid buffer, for example we did alltoallv with 64bit displacement in pytorch

Reply:
you are right. still worth moving to separate fn imho, even with 2 more flags. current function is very large. just from readability perspective.